src/p2p/net_node.inl

```diff
@@ -155,18 +155,34 @@ namespace nodetool
   }
   //-----------------------------------------------------------------------------------
   template<class t_payload_net_handler>
-  bool node_server<t_payload_net_handler>::is_remote_host_allowed(const epee::net_utils::network_address &address)
+  bool node_server<t_payload_net_handler>::is_remote_host_allowed(const epee::net_utils::network_address &address, time_t *t)
   {
     CRITICAL_REGION_LOCAL(m_blocked_hosts_lock);
-    auto it = m_blocked_hosts.find(address.host_str());
-    if(it == m_blocked_hosts.end())
-      return true;
-    if(time(nullptr) >= it->second)
+
+    // manually loop since we have to check for subnets
+    const time_t now = time(nullptr);
+    std::map<epee::net_utils::network_address, time_t>::iterator it;
+    for (it = m_blocked_hosts.begin(); it != m_blocked_hosts.end(); )
```

Comment:
Why have a `std::map` if find operations no longer work?

A `ipv4_network_address` instance should be merged into a `ipv4_network_subnet` instance such that the `find` operation still works. Turning a logarithmic operation into a linear one seems undesirable in this context, although hopefully most nodes have a low blocked hosts count.

`boost::variant<ipv4_network_subnet, std::string>` would work with minimal effort. Or a "stripped down" copy-and-paste of `network_address` that has special assignment for `ipv4_network_address` that merges into `ipv4_network_subnet` would work too. Only serialization, `operator<` and `str()` methods need to be copied really since there is no such concept of port with this type.

Reply:
No real reason. It could be made a deque instead.

You'll have to elaborate about your last paragraph, I'm not sure what you're suggesting.

Reply:
> You'll have to elaborate about your last paragraph, I'm not sure what you're suggesting.

`boost::variant` has an `operator<` that sorts by element type first so it can be used in a `std::map` with `find`:

```c++
std::map<
  boost::variant<epee::net_utils::ipv4_network_subnet, std::string>, time_t
> m_blocked_hosts;
...

bool is_remote_host_allowed(const epee::net_utils::network_address &address, time_t *t)
{
  using epee::net_utils;

  boost::variant<ipv4_network_subnet, std::string> host;
  switch (address.get_type_id())
  {
    case address_type::ipv4:
      host = ipv4_network_subnet{address.as<ipv4_network_address>().ip(), 32};
      break;
    default:
      host = address.host_str();
      break;
  }
  auto it = m_blocked_hosts.find(host);
}
```
which would keep the logarithmic lookup intact. One small benefit over the current mainline code is that when `address` is an ipv4 address, no temporary memory allocations need to be made anymore.

The other technique I suggested (copying `network_address`) would be a little more work than I first thought since existing network address types implemented `operator<` and `operator==` in terms of host AND port. 

And so I think this patch changed lookup behavior slightly with `operator==` usage instead of `find(address.host_str())`. The issue would only occur with non IPv4 addresses.

Reply:
OK, that looks alright. Any reason for using std::string rather than network_address, beyond it is the current way ? Most places will just call host_str, then lookup with that, it seems like a waste.

Reply:
I used `std::string` in that example because it was the easiest technique to update the existing implementation. The problem is that `operator==` and `operator<` use exact matching with port whereas comparing `host_str()` output only compares the host.

Reply:
Re-reading your code, I think it doesn't work in this case:
m_blocked_hosts contains 1.2.3.0/24
Queried address is 1.2.3.4
Then host gets set to 1.2.3.4/32, which should be blocked, but isn't.

Reply:
Changing from a log search to a linear search is still unfortunate. "Stuffing" this table should be tough remotely, although I haven't thought about it thoroughly.

And the above code should have been `m_blocked_hosts.lower_bound(host);`. And the other issues:
  * The value in `ipv4_network_subnet` is network order but needs to be host order for proper searching
  * A search must be done against an `operator<(subnet, ip)` so IP to subnet is converted properly. This requires one of:
    * Moving to C++14
    * Updating minimum boost version
    * Using a sorted `std::vector` instead